File: orcastork/adapters/memory/rate_limiter.py
"""In-memory ``RateLimiter`` — a classic token bucket per key on the injected clock.

Single-process only (the Redis adapter coordinates across pods); deterministic under a
``FakeClock`` because both the refill arithmetic (``monotonic``) and the wait (``sleep``)
go through the injected clock.
"""

from __future__ import annotations

from dataclasses import dataclass

from ...clock import Clock


@dataclass
class _Bucket:
    tokens: float
    refilled_at: float  # monotonic time of the last refill computation


class InMemoryRateLimiter:
    def __init__(self, clock: Clock, *, rate_per_second: float, burst: int) -> None:
        # Fail fast on a misconfigured limiter rather than dividing by zero (or never refilling) on
        # the first contended acquire.
        if rate_per_second <= 0.0:
            raise ValueError(f'rate_per_second must be > 0.0, got {rate_per_second}')
        if burst < 1:
            raise ValueError(f'burst must be >= 1, got {burst}')
        self._clock = clock
        self._rate = rate_per_second
        self._burst = burst
        self._buckets: dict[str, _Bucket] = {}

    async def acquire(self, key: str) -> None:
        # Re-check after every sleep: a concurrent waiter may have taken the token this one slept for.
        while True:
            now = self._clock.monotonic()
            bucket = self._buckets.setdefault(key, _Bucket(tokens=float(self._burst), refilled_at=now))
            bucket.tokens = min(float(self._burst), bucket.tokens + (now - bucket.refilled_at) * self._rate)
            bucket.refilled_at = now
            if bucket.tokens >= 1.0:
                bucket.tokens -= 1.0
                return
            await self._clock.sleep((1.0 - bucket.tokens) / self._rate)

File: typescript/src/orcastork/adapters/memory/rate_limiter.ts
/**
 * In-memory `RateLimiter` — a classic token bucket per key on the injected clock.
 *
 * Single-process only (the Redis adapter coordinates across pods); deterministic under a
 * `FakeClock` because both the refill arithmetic (`monotonic`) and the wait (`sleep`) go through
 * the injected clock.
 *
 * @module
 */

import { z } from 'zod';
import type { Clock } from '../../clock.js';
import type { RateLimiter } from '../../ports/rate_limiter.js';

/** Milliseconds in the second `ratePerSecond` is expressed in. */
const MS_PER_SECOND = 1000;

/**
 * The bounds a token bucket must satisfy, whichever backend runs it.
 *
 * Exported so the Redis limiter rejects the same misconfiguration identically: the two adapters
 * share the "waits, never fails" contract, and a zero/negative rate would never refill (hang)
 * while a sub-1 burst would divide by zero on the first contended acquire.
 */
export const tokenBucketBounds = {
  ratePerSecond: z.number().positive(),
  burst: z.number().int().min(1),
};

const tokenBucketSchema = z.object(tokenBucketBounds);

/** How a token bucket is paced: its steady-state rate and how much of it may be spent at once. */
export interface TokenBucketOptions {
  /** Tokens accrued per second; the steady-state pace of the key. */
  readonly ratePerSecond: number;

  /** Bucket capacity — the most actions that may fire back to back after an idle period. */
  readonly burst: number;
}

/** One key's bucket: how many tokens it holds and when that was last computed. */
interface Bucket {
  tokens: number;

  /** Monotonic time (ms) of the last refill computation. */
  refilledAt: number;
}

/** Fleet pacing for a single process — the test and local-run substrate. */
export class InMemoryRateLimiter implements RateLimiter {
  private readonly clock: Clock;
  private readonly ratePerSecond: number;
  private readonly burst: number;
  private readonly buckets = new Map<string, Bucket>();

  public constructor(clock: Clock, options: TokenBucketOptions) {
    // Fail fast on a misconfigured limiter rather than dividing by zero (or never refilling) on the
    // first contended acquire.
    const validated = tokenBucketSchema.parse(options);
    this.clock = clock;
    this.ratePerSecond = validated.ratePerSecond;
    this.burst = validated.burst;
  }

  public async acquire(key: string): Promise<void> {
    // Re-check after every sleep: a concurrent waiter may have taken the token this one slept for.
    for (;;) {
      const now = this.clock.monotonic();
      const bucket = this.bucketFor(key, now);
      const accrued = ((now - bucket.refilledAt) / MS_PER_SECOND) * this.ratePerSecond;
      bucket.tokens = Math.min(this.burst, bucket.tokens + accrued);
      bucket.refilledAt = now;
      if (bucket.tokens >= 1) {
        bucket.tokens -= 1;
        return;
      }
      await this.clock.sleep(((1 - bucket.tokens) / this.ratePerSecond) * MS_PER_SECOND);
    }
  }

  private bucketFor(key: string, now: number): Bucket {
    const known = this.buckets.get(key);
    if (known !== undefined) {
      return known;
    }
    const created: Bucket = { tokens: this.burst, refilledAt: now };
    this.buckets.set(key, created);
    return created;
  }
}
